packages/iov-encoding/src/decimal.ts

```diff
@@ -0,0 +1,86 @@
+import BN from "bn.js";
+
+/**
+ * A type for arbitrary precision, non-negative decimals.
+ *
+ * Instances of this class are immutable.
+ */
+export class Decimal {
+  public static fromUserInput(input: string, fractionalDigits: number): Decimal {
+    if (!Number.isInteger(fractionalDigits)) throw new Error("Fractional digits is not an integer");
+    if (fractionalDigits < 0) throw new Error("Fractional digits must not be negative");
+
+    for (let index = 0; index < input.length; index++) {
```

Comment:
It would be cleaner and faster to use a negated set:
```ts
const badCharacter = input.match(/[^\d\.]/);
if (badCharacter) {
    throw new Error(`Invalid character at position ${badCharacter.index + 1}`);
}
```